docs/userguide/mpi_apps.rst

```diff
@@ -40,116 +73,275 @@ to the desired number of blocks.
+In MPI mode, the :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` can execute both Python or Bash Apps which invokes the MPI application.
+However, it is important to not that Python Apps that directly use ``mpi4py`` is not supported.
 
+For multi-node MPI applications, especially when running multiple applications within a single batch job,
+it is important to specify the resource requirements for the app so that the Parsl worker can provision
+the appropriate resources before the application starts. For eg, your Parsl script might contain a molecular
+dynamics application that requires 8 ranks over 1 node for certain inputs and 32 ranks over 4 nodes for some
+depending on the size of the molecules being simulated. By specifying resources via ``resource_specification``,
+parsl workers will provision the requested resources and then compose MPI launch command prefixes
+(Eg: ``mpiexec -n <ranks> -ppn <ranks_per_node> -hosts <node1..nodeN>``). These launch command prefixes are
+shared with the app via environment variables.
 
 .. code-block:: python
 
     @bash_app
-    def echo_hello(n: int, stderr='std.err', stdout='std.out'):
-        return f'mpiexec -n {n} --ppn 1 hostname'
+    def echo_hello(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
+        return f'$PARSL_MPI_PREFIX hostname'
+
+    # Alternatively, you could also use the resource_specification to compose a launch
+    # command using env vars set by Parsl from the resource_specification like this:
+    @bash_app
+    def echo_something(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
+        total_ranks = os.environ("")
+        return f'aprun -N $PARSL_RANKS_PER_NODE -n hostname'
+
 
-Complications arise when running more than one MPI task per block,
-and the MPI launcher does not automatically spread jobs across nodes.
-In this case, use the ``PARSL_WORKER_RANK`` environment variable
-set by HTEx to select the correct hostfile:
+All key-value pairs set in the resource_specification are exported to the application via env vars, for eg.
+``parsl_resource_specification = {'MY_ENV_VAR': 'MY_ENV_VALUE'} `` will set the env var ``MY_ENV_VAR`` to
+``MY_ENV_VALUE`` in the application's env.
 
+However, the following options are **required** for MPI applications :
 
 .. code-block:: python
 
-    @bash_app
-    def echo_hello(n: int, stderr='std.err', stdout='std.out'):
-        return (f'mpiexec -n {n} --ppn 1 '
-                '--hostfile /tmp/hostfiles/local_hostfile.`printf %02d $PARSL_WORKER_RANK` '
-                'hostname')
+    resource_specification = {
+      'NUM_NODES': <int>,        # Number of nodes required for the application instance
+      'RANKS_PER_NODE': <int>,   # Number of Ranks / application elements to be launched per node
+    }
 
-.. note::
+    # The above are made available in the worker env vars:
+    # echo $PARSL_NUM_NODES, $PARSL_RANKS_PER_NODE
```

Comment:
Is there an environment variable for the list of nodes selected by Parsl? I'll need that to use MPI mode when launching deepspeed tasks

Reply:
Yep, I've added additional info on the env vars we support, in this case `PARSL_MPI_NODELIST` is what you are looking for.

Reply:
Woo!